### PR TITLE
Add distribute_mg_dof functionality for a parallel::shared::Triangulation

### PR DESCRIPTION
**Pull request by tcclevenger**
- Distribute mg dofs for a parallel::shared::Triangulation
- Distributes level dofs in the same way as distributed triangulation (proc with smallest index gets boundary dofs)
- MG computations can now be run on an adaptive mesh using shared triangulation 
- Added 2 tests for checking dof distribution and 1 test which is a working version of step-50 for a shared triangulation

### CODE REVIEW

**Comment by tcclevenger** [comment]
I need to update my branch. I'll create a new pull request when I'm finished

**Comment by davydden** [comment]
@tcclevenger no need for a new PR. just do
```
git checkout master
git pull
git checkout sharedtria_mg_dofs
git rebase master
// check that rebase is good, things still work
git push -f
```

**Comment by tcclevenger** [comment]
I looked at where the build failed. It has to do with the function `shared::Triangulation::get_true_level_subdomain_ids_of_cells` I am a bit confused since I recompiled on my laptop and there was no error. All of my tests also passed. I was getting a similar compile error earlier but I fixed mine with the `added template commit`. It started to build before I sent in that commit, but I'm sure I could be wrong on when the build is executed.

**Comment by masterleinad** [comment]
I can compile this without problems. Hence, I expect tjehi-gcc-serial to compile successfully after your next change to this branch.

**Comment by tjhei** [comment]
No. That configuration is without MPI. Note that your function only exists with MPI (the whole shared Tria stuff is inside an ``#ifdef``), but we are still compiling the policy .cc. You probably need to wrap the call with an ifdef.

**Comment by tcclevenger** [comment]
This makes sense to me, but is there a good explanation for why `shared::Triangulation::get_true_subdomain_ids` is defined in the same way and place in `shared_tria.h/.cc` and used without an `#ifdef` in `dof_handler_policy.cc` function `distribute_dofs()`?
The only difference I can see is that `distribute_dofs` has `if (tr->with_artificial_cells())` before it calls `get_true_subdomain_ids` whereas `distribute_mg_dofs` has no such if-statement (sharedtrai enforces `with_artificial_cells` for mg computation), but this shouldn't matter for a compiling error, right?

**Comment by tcclevenger** [comment]
Sorry for all the commits. I don't have a version of dealii compiled without MPI to check against (which I'll set up now in case this happens again). Everything is working now. Again:
- Distribute mg dofs for a parallel::shared::Triangulation
- Distributes level dofs in the same way as distributed triangulation (proc with smallest index gets boundary dofs)
- MG computations can now be run on an adaptive mesh using shared triangulation with zorder partitioning or user defined partition (as long as level_subdomain_id=subdomain_id for active cells)
- Added 2 tests for checking dof distribution and 1 test which is a working version of step-50 for a shared triangulation

**Comment by bangerth** [comment]
What's the status of this? I think we have collectively lost track of it.

**Comment by tcclevenger** [comment]
It's still at its beginning stage. I was having issues passing the check for compiling without MPI (all technicalities since sharedtria will never be run without MPI=on) but it is now corrected.  Treat my previous comment as the initial comment for the pull request.

**Comment by tjhei** [comment]
I think what @tcclevenger is trying to say is that it is done and ready for review/merge.

File: source/distributed/shared_tria.cc [hunk 1]
```diff
@@ -274,6 +290,83 @@ namespace parallel
...
     }
 
+
+
+    template <int dim, int spacedim>
+    void
+    Triangulation<dim,spacedim>::
+    update_number_cache ()
+    {
+      parallel::Triangulation<dim,spacedim>::update_number_cache();
```
Comment (tjhei) [inline source/distributed/shared_tria.cc]: this is the same implementation as p::d::Tria, right? Just move it into p::Tria and delete this and the function update_number_cache() in p::d::Tria.

File: source/distributed/tria_base.cc [hunk 2]
```diff
@@ -191,11 +191,78 @@ namespace parallel
...
-        if (cell->subdomain_id() == my_subdomain)
-          ++number_cache.n_locally_owned_active_cells[my_subdomain];
+      {
+        for (typename Triangulation<dim,spacedim>::active_cell_iterator
+             cell = this->begin_active();
+             cell != this->end(); ++cell)
+          if (cell->subdomain_id() == my_subdomain)
+            ++number_cache.n_locally_owned_active_cells[my_subdomain];
+
+        if (1)//settings & construct_multigrid_hierarchy)
```
Comment (tcclevenger) [inline source/distributed/tria_base.cc]: Obviously this needs to be changed. We only want to fill level_ghost_owners if we are running a MG computation, but tria_base.h doesn't have this setting. The only settings variable that both p::d::Tria and p::s::Tria share is the construct_multigrid_hierarchy. Is there a way that I could ask the triangulation here to tell me its settings? Or will I need to give the base class a member bool variable that stores whether or not this setting is turned on, and set it in p::d::T and p::s::T if the setting is turned on?

Comment (tjhei) [reply source/distributed/tria_base.cc]: That is annoying. No, sadly the settings are declared in ``p::d::Tria`` and ``p::s::Tria`` separately. Without doing large restructuring, I would do the following:
1. create a ``protected`` function ``fill_level_ghost_owners()`` inside the base class.
2. re-add the ``update_number_cache`` in the derived classes to do something like:
```
parallel::Triangulation<dim,spacedim>::update_number_cache();
if (settings & construct_multigrid_hierarchy)
  fill_level_ghost_owners();
```
I hope that works...

**Comment by tcclevenger** [comment]
`fill_level_ghost_owners` has been added to `tria_base` and is called from `p::d::T` and `p::s::T` in their respective `update_number_cache` functions.

**Comment by tjhei** [comment]
Okay, can you please squash your commits? I think this looks good now.

**Comment by tjhei** [comment]
This is now ready (in my opinion) if somebody else wants to take a look.

File: source/distributed/shared_tria.cc (hunk 1 above)
Comment (tjhei) [reply source/distributed/shared_tria.cc]: No. We can not move the function to the base class, as it doesn't know about ``settings & construct_multigrid_hierarchy`` (the settings are independently defined in both derived classes). He did move the duplicated code into the new ``fill_level_ghost_owners``.

Comment (bangerth) [reply source/distributed/shared_tria.cc]: Ah, ok. Thanks for confirming.

**Comment by tjhei** [comment]
Can you write a changelog entry (I think this is big enough to be under "major")?

**Comment by tcclevenger** [comment]
I've made the requested changes and added the changelog entry. Let me know if there is anything else we need here. Thanks!

**Comment by bangerth** [comment]
Nope, that's all -- thank you very much for the contribution!

**Comment by tcclevenger** [comment]
I am not exactly sure why the one test failed since it seems to have nothing to do with a `parallel::Triangulation`. The information was given

The following tests FAILED:
	1622 - grid/intergrid_constraints.debug (Timeout)
Errors while running CTest

I just ran it on my own computer and the test passed, but this was not with a serial build of dealii.

**Comment by drwells** [comment]
That test was recently written to be less troublesome; see #5203. Don't worry about it.

**Comment by bangerth** [comment]
Yes, no worries -- we know the test occasionally times out, and @tjhei has already had his patch in #5203 merged in which the test is shortened.

**Comment by bangerth** [comment]
I had meant to merge this earlier today already, but apparently forgot to hit the "confirm" button.

**Pull request merged**